lgc/patch/PatchEntryPointMutate.cpp

```diff
@@ -1325,6 +1281,69 @@ void PatchEntryPointMutate::addUserDataArg(SmallVectorImpl<UserDataArg> &userDat
+// Add a PushConst UserDataArg
+//
+// @param userDataArgs : Vector to add UserDataArg to
+// @param node : PushConst resource node
+// @param builder : IRBuilder (just for getting types)
+void PatchEntryPointMutate::addPushConstArgs(SmallVectorImpl<UserDataArg> &userDataArgs, const ResourceNode &node,
+                                             llvm::IRBuilder<> &builder) {
+  auto userDataUsage = getUserDataUsage(m_shaderStage);
+  // Always spill for compute libraries.
+  if (isComputeWithCalls()) {
+    // Mark push constant for spill for compute library.
+    userDataUsage->pushConstSpill = true;
+  } else {
+    // We add a potential unspilled arg for each separate dword offset of the push const at which there is a load.
+    // We already know that loads we have on our pushConstOffsets lists are at dword-aligned offset and
+    // dword-aligned size. We need to ensure that all loads are the same size, by removing ones that are bigger than
+    // the minimum size.
+    //
+    // First cope with the case that the app uses more push const than the size of the resource node. This is
+    // a workaround for an incorrect application; according to the Vulkan spec (version 1.2.151, section 14.6.1
+    // "Push Constant Interface"):
+    //
+    //    Each statically used member of a push constant block must be placed at an Offset such that the entire
+    //    member is entirely contained within the VkPushConstantRange for each OpEntryPoint that uses it, and
+    //    the stageFlags for that range must specify the appropriate VkShaderStageFlagBits for that stage.
+    unsigned dwordEndOffset = userDataUsage->pushConstOffsets.size();
+    if (dwordEndOffset > node.sizeInDwords) {
+      userDataUsage->pushConstSpill = true;
+      dwordEndOffset = node.sizeInDwords;
+    }
+
+    for (unsigned dwordOffset = 0; dwordOffset != dwordEndOffset; ++dwordOffset) {
+      UserDataNodeUsage &pushConstOffset = userDataUsage->pushConstOffsets[dwordOffset];
+      if (pushConstOffset.users.empty())
+        continue;
+
+      // Check that the load size does not overlap with the next used offset in the push constant.
+      bool haveOverlap = false;
+      unsigned endOffset =
+          std::min(dwordOffset + pushConstOffset.dwordSize, unsigned(userDataUsage->pushConstOffsets.size()));
+      for (unsigned followingOffset = dwordOffset + 1; followingOffset != endOffset; ++followingOffset) {
+        if (!userDataUsage->pushConstOffsets[followingOffset].users.empty()) {
+          haveOverlap = true;
+          break;
+        }
+      }
+      if (haveOverlap) {
+        userDataUsage->pushConstSpill = true;
+        continue;
+      }
+
+      // Add the arg (part of the push const) that we can potentially unspill.
+      addUserDataArg(userDataArgs, node.offsetInDwords + dwordOffset, pushConstOffset.dwordSize,
+                     "pushConst" + Twine(dwordOffset), &pushConstOffset.entryArgIdx, builder);
```

Comment:
I would expect that in the indirect case, these values will always have to be unspilled.  Can you point out the code that will do the unspilling?

Reply:
Sorry, I think I meant that it should never be unspilled.  Are we sure that will not happen?

Reply:
The code is the same as before. I only move it. I think it should be correct, after all it has been stable. Thanks.

Reply:
Even if this is the same code that does not mean it is correct because you are executing it in a different context.  For example, should you be setting `userDataUsage->spillUsage` below if the push constant node is not in the root table?

I still wonder what will happen with the pushConst argument in `PatchEntryPointMutate::determineUnspilledUserDataArgs`.  Again, that code was written assuming that push constants are in the root table.  What will that code do if it decides pass the push constant in SGPRs instead of loading it from memory?  I assume that should never happen, but no code was added to make sure it does not.